content/4_Gold/Shortest_Paths.mdx

```diff
@@ -66,6 +64,113 @@ idea to learn Bellman-Ford and Floyd-Warshall first since they're simpler.
 
 </Optional>
 
+### Problem
+
+<FocusProblem problem="apspSam" />
+
+#### Explanation
+
+This problem asks us to compute shortest paths between any two vertices. Hence,
+Floyd-Warshall is suitable because of the low $N (N \le 500)$, and the inclusion of negative weights.
+
+#### Implementation
+
+**Time Complexity:** $\mathcal{O}(N^3)$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+//CodeSnip{CPP Short Template}
+
+ll INF = 1e18;
+
+int main() {
+	ios::sync_with_stdio(0);
+	cin.tie(0);
+	int n, m, q;
+	cin >> n >> m >> q;
+
+	vector<vector<ll>> mat(n, vector<ll>(n, INF));
+
+	for (int i = 0; i < m; i++) {
+		int u, v, w;
+		cin >> u >> v >> w, u--, v--;
+		if (w < mat[u][v]) mat[u][v] = mat[v][u] = w;
+	}
+
+	// floyd-warshall
+	for (int k = 0; k < n; k++)
+		for (int i = 0; i < n; i++)
+			for (int j = i + 1; j < n; j++)
+				if (mat[i][k] + mat[k][j] < mat[i][j])
+					mat[i][j] = mat[j][i] = mat[i][k] + mat[k][j];
+
+	while (q--) {
+		int u, v;
+		cin >> u >> v, u--, v--;
+		if (u == v) mat[u][v] = 0;
+		if (mat[u][v] == INF) mat[u][v] = -1;
+		cout << mat[u][v] << "\n";
+	}
+}
+```
+
+</CPPSection>
+
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+
+public class Main {
+	// CodeSnip{Kattio}
+
+	public static void main(String[] args) {
+		Kattio io = new Kattio();
+
+		int n = io.nextInt(), m = io.nextInt(), q = io.nextInt();
+
+		long[][] mat = new long[n][n];
+		long INF = (long)1e18;
```

Comment:
Why not just Long.MAX_VALUE

Reply:
its good practice to set inf as a little below the max value so you dont overflow.

Reply:
wouldn't you be doing something wrong if you're adding to INF anyways?